### PR TITLE
Teacher proxy

### PR DESCRIPTION
**Pull request by andrzejgorski**

### CODE REVIEW

File: whylog_vim/proxy/teacher/__init__.py [hunk 1]
```diff
@@ -1,17 +1,60 @@
...
     def __init__(self, teacher, editor, main_proxy):
         self.teacher = teacher
         self.editor = editor
         self.main_proxy = main_proxy
 
+    def new_lesson(self):
+        front_input = self.editor.get_front_input()
+        self.teacher.add_line(get_next_parser_id(), front_input, effect=True)
+        self.origin_file_name = self.editor.get_current_filename()
+        print Messages.ADDED_EFFECT
```
Comment (kgromadzki) [inline whylog_vim/proxy/teacher/__init__.py]: Remove this print

File: whylog_vim/proxy/teacher/__init__.py [hunk 2]
```diff
@@ -1,17 +1,60 @@
...
     def __init__(self, teacher, editor, main_proxy):
         self.teacher = teacher
         self.editor = editor
         self.main_proxy = main_proxy
 
+    def new_lesson(self):
+        front_input = self.editor.get_front_input()
+        self.teacher.add_line(get_next_parser_id(), front_input, effect=True)
+        self.origin_file_name = self.editor.get_current_filename()
+        print(Messages.ADDED_EFFECT)
```
Comment (andrzejgorski) [inline whylog_vim/proxy/teacher/__init__.py]: Print is the standard method to send message to user in python-vim.

File: whylog_vim/proxy/teacher/utils.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from functools import partial
+
+
+def naturals_generator():
```
Comment (kgromadzki) [inline whylog_vim/proxy/teacher/utils.py]: itertools.count(0) ?

File: whylog_vim/proxy/teacher/utils.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from functools import partial
+from itertools import count
+
+parser_ids = count(0)
```
Comment (kgromadzki) [inline whylog_vim/proxy/teacher/utils.py]: Refactor to:

``` python
get_next_parser_id = partial(next, count())
get_next_constraints_id = partial(next, count())
```

File: whylog_vim/proxy/teacher/__init__.py (hunk 2 above)
Comment (ppolewicz) [inline whylog_vim/proxy/teacher/__init__.py]: don't you need to `from __future__ import print_statement` to use it under python2?

File: whylog_vim/proxy/teacher/__init__.py [hunk 5]
```diff
@@ -1,17 +1,60 @@
...
+
+    def _return_cursor_to_position(self):
+        try:
+            self.editor.go_to_offset(self._return_offset)
+        except Exception:
+            raise CannotGoToPosition(self._return_offset)
+        else:
+            self.editor.open_fold()
+
+    def set_return_function(self, return_function):
```
Comment (ppolewicz) [inline whylog_vim/proxy/teacher/__init__.py]: redefinition of method from line 31

File: whylog_vim/proxy/teacher/__init__.py [hunk 6]
```diff
@@ -1,17 +1,60 @@
...
+        self.print_teacher()
+
     def handle_menu_signal(self):
-        pass
+        self._set_cursor_position()
+        self.output.call_button(self.editor.get_line_number())
+        if self.main_proxy.get_state() == EditorStates.TEACHER:
+            self._return_cursor_to_position()
+
+    def set_read_function(self, read_function):
```
Comment (ppolewicz) [inline whylog_vim/proxy/teacher/__init__.py]: this method is essentially a setter. You don't need to define those in python, but can use `object.field = value` instead and make it a setter later if necessary.

From reading this class it is hard to understand what the `read_function` is for. Please document it.

File: whylog_vim/proxy/teacher/__init__.py [hunk 7]
```diff
@@ -1,17 +1,60 @@
...
 
-    def add_cause(self):
-        pass
+    def _set_cursor_position(self):
+        self._return_offset = self.editor.get_offset()
+
+    def _return_cursor_to_position(self):
+        try:
+            self.editor.go_to_offset(self._return_offset)
+        except Exception:
```
Comment (ppolewicz) [inline whylog_vim/proxy/teacher/__init__.py]: too wide Exception catching. Please catch more specific exception types.

File: whylog_vim/tests/tests_proxy/tests_unit_proxy_main.py [hunk 8]
```diff
@@ -1,25 +1,31 @@
...
 from whylog_vim.proxy import WhylogProxy
 
 
 class UnitTestWhylogProxy(TestCase):
-    def tests_unit_check_log_reader_states_of_whylog_proxy(self):
-        whylog_proxy = WhylogProxy(MagicMock())
+    def setUp(self):
+        mock_editor = MagicMock()
+        mock_editor.get_front_input.return_value = FrontInput(
+                1, 'line content', LineSource('host', 'path'))  # yapf:disable
```
Comment (ppolewicz) [inline whylog_vim/tests/tests_proxy/tests_unit_proxy_main.py]: don't disable yapf here and even when you do disable it, format the line properly before disabling it

**Comment by ppolewicz** [comment]
merge conflict

**Comment by ppolewicz** [comment]
CI failures - please fix and assign back to me

**Pull request merged**